### PR TITLE
deadlock in system.text_log

### LINKED ISSUE #12325: Deadlock in system.text_log
**Describe the bug**
We add a log message to the `system.text_log` table with mutex https://github.com/ClickHouse/ClickHouse/blob/master/base/loggers/OwnSplitChannel.cpp#L95-L97. But surprisingly in `log->add` method we also have logging: here https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/SystemLog.h#L253 or from memory tracker. This logging can lead to deadlock because the same thread will try to acquire the same mutex one more time. After this server became irresponsible.
Example trace of such thread:
```
Thread 1235 (Thread 0x7f6d475fb700 (LWP 166244)):
#0  0x00007f6d5e6f1f1c in __lll_lock_wait () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f6d5e6ed649 in _L_lock_909 () from /lib/x86_64-linux-gnu/libpthread.so.0
#2  0x00007f6d5e6ed470 in pthread_mutex_lock () from /lib/x86_64-linux-gnu/libpthread.so.0
#3  0x0000000011088d4d in std::__1::mutex::lock() ()
#4  0x000000000917f521 in DB::OwnSplitChannel::logSplit(Poco::Message const&) ()
#5  0x0000000009181443 in DB::OwnSplitChannel::log(Poco::Message const&) ()
#6  0x0000000008f4f255 in MemoryTracker::alloc(long) ()
#7  0x0000000008f4eec3 in MemoryTracker::alloc(long) ()
#8  0x0000000008ef8cec in operator new(unsigned long) ()
#9  0x000000001108dafc in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::basic_string(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) ()
#10 0x00000000091828f1 in DB::SystemLog<DB::TextLogElement>::add(DB::TextLogElement const&) ()
#11 0x000000000917f550 in DB::OwnSplitChannel::logSplit(Poco::Message const&) ()
#12 0x0000000009181443 in DB::OwnSplitChannel::log(Poco::Message const&) ()
#13 0x000000000dab5cef in DB::ReplicatedMergeTreeQueue::shouldExecuteLogEntry(DB::ReplicatedMergeTreeLogEntry const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, DB::MergeTreeDataMergerMutator&, DB::MergeTreeData&, std::__1::lock_guard<std::__1::mutex>&) const ()
#14 0x000000000dabe6d7 in DB::ReplicatedMergeTreeQueue::selectEntryToProcess(DB::MergeTreeDataMergerMutator&, DB::MergeTreeData&) ()
#15 0x000000000d728459 in DB::StorageReplicatedMergeTree::queueTask() ()
#16 0x000000000d8878bb in DB::BackgroundProcessingPool::threadFunction() ()
#17 0x000000000d8882b0 in ?? ()
#18 0x0000000008f757e7 in ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) ()
#19 0x0000000008f73c33 in ?? ()
#20 0x00007f6d5e6eb184 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#21 0x00007f6d5ef0937d in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Try to avoid deadlock in `system.text_log`. Closes #12325

### CODE REVIEW

**Comment by akuzm** [comment]
It may be interesting to add a scoped class like `getCurrentMemoryTrackerActionLock` that disables memory allocation at all, i.e. makes all new() calls throw. This way we can be sure that there is no allocation.

File: src/Interpreters/SystemLog.h [hunk 1]
```diff
@@ -272,7 +276,7 @@ void SystemLog<LogElement>::add(const LogElement & element)
         return;
     }
 
-    queue.push_back(element);
+    queue.push_back(std::forward<T>(element));
```
Comment (akuzm) [inline src/Interpreters/SystemLog.h:279]: This should avoid copy-construction where possible, but not always -- depends on what kind of reference you pass. Maybe we should require rvalue reference input instead?

Also there is still reallocation of `queue` buffer itself -- we can probably grow it to max size right away. If you do this, don't forget also initialize the local queue in the flushing thread, that is swapped with this one.

Comment (nikitamikhaylov) [reply src/Interpreters/SystemLog.h:279]: 1) I thought the problem was only with `text_log`. Other logs will pass lvalue as before. 
2) That is the problem.

**Comment by nikitamikhaylov** [comment]
@akuzm I forgot why we have to avoid allocations in `log` function? The memory tracker is already off via `getCurrentMemoryTrackerActionLock`. Also it is not possible to use `std::move` everywhere, so I returned back universal link with template parameter. And added a function to test it somehow. It seems working.

File: src/Functions/printToLog.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+class FunctionPrintToLog : public IFunction
+{
+public:
+    static constexpr auto name = "printToLog";
```
Comment (akuzm) [inline src/Functions/printToLog.cpp:23]: Can I now run `select printToLog(repeat('a', 1000)) from system.numbers` to bring down the playground? It should require sysadmin priveleges and be disabled in read only mode.

File: base/loggers/OwnSplitChannel.cpp [hunk 3]
```diff
@@ -92,9 +92,13 @@ void OwnSplitChannel::logSplit(const Poco::Message & msg)
...
 
-        std::lock_guard<std::mutex> lock(text_log_mutex);
+        std::unique_lock<std::mutex> lock(text_log_mutex);
         if (auto log = text_log.lock())
-            log->add(elem);
+        {
+            lock.unlock();
+            log->add(std::move(elem));
+        }
+
```
Comment (akuzm) [inline base/loggers/OwnSplitChannel.cpp:101]: I can already imagine how I fail to notice the `lock.unlock()` inside the `if`, add some code after `if` and believe that it's under the lock. Let's restructure:
```
shared_ptr<TextLog> text_log_locked;
{
    std::lock_guard<std::mutex> lock(text_log_mutex);
    text_log_locked = text_log.lock();
}
if (text_log_locked)
{
   text_log_locked->add(elem);
}
```

**Comment by akuzm** [comment]
> @akuzm I forgot why we have to avoid allocations in `log` function? The memory tracker is already off via `getCurrentMemoryTrackerActionLock`. Also it is not possible to use `std::move` everywhere, so I returned back universal link with template parameter. And added a function to test it somehow. It seems working.

Idk, why do we have to use `std::move` then? Maybe we should leave only the locking changes in OwnSplitChannel?

**Comment by akuzm** [comment]
The new function is probably not something we should backport. So I guess we should have a one-line PR with locking changes and bugfix category, and everything else in another PR. Or do we have to backport it at all? I remember Alexey fixed something, maybe it's enough for the back branches?

**Comment by akuzm** [comment]
> Idk, why do we have to use `std::move` then? Maybe we should leave only the locking changes in OwnSplitChannel?

But I think the changes themselves are OK and we can keep them. These logs can be used in surprisingly intensive ways e.g. when tracing all memory allocations -- I had to fix several performance problems before. So saving excessive data copying and allocation is helpful.

File: src/Functions/printToLog.cpp (hunk 2 above)
Comment (alexey-milovidov) [reply src/Functions/printToLog.cpp:23]: `allow_introspection_functions`

Comment (alexey-milovidov) [reply src/Functions/printToLog.cpp:23]: I also wanted this function, see https://github.com/ClickHouse/ClickHouse/issues/9434

Details:
- only trace level;
- no quantity parameter (I think it's harmful);
- different name (log is also a verb);

File: src/Functions/printToLog.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    bool isDeterministic() const override { return true; }
+    bool isDeterministicInScopeOfQuery() const override { return true; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes &) const override
+    {
+        return std::make_shared<DataTypeString>();
```
Comment (alexey-milovidov) [inline src/Functions/printToLog.cpp:42]: I'd return 0 of type UInt8.

**Comment by akuzm** [comment]
The deadlock is fixed.
The std::move changes I'm not sure are needed at all.

What's left is the print function. @nikitamikhaylov do you want to continue? Maybe we should close this PR and concentrate on something more vital.

**Comment by nikitamikhaylov** [comment]
> The deadlock is fixed.
> 
> The std::move changes I'm not sure are needed at all.
> 
> 
> 
> What's left is the print function. @nikitamikhaylov do you want to continue? Maybe we should close this PR and concentrate on something more vital.

Of course, maybe I will create PR only with function in the near future.

**Comment by akuzm** [comment]
> Of course, maybe I will create PR only with function in the near future.

OK, I'll close this one then.

**Pull request closed without merging**